File: ruby/maximum-subarray.rb
#!/usr/bin/env ruby

# Find the contiguous subarray within an array (containing at least one number) which has the largest sum.

# For example, given the array [-2,1,-3,4,-1,2,1,-5,4],
# the contiguous subarray [4,-1,2,1] has the largest sum = 6.

# @param {Integer[]} nums
# @return {Integer}
def max_sub_array(nums)
  return 0 if nums.nil? or nums.empty?
  size = nums.size
  dp = Array.new(size + 1, 0)
  max = nums[0]
  0.upto(size - 1) do |i|
    i.upto(size - 1) do |j|
      dp[j + 1] = nums[j] + dp[j] if i == 0
      res = dp[j + 1] - dp[i]
      max = res if res > max
    end
  end
  max
end

nums_string = ARGV[0]
nums = nums_string[1...-1].split(",").map{|x| x.to_i}
puts "nums are #{nums}"
puts "res is #{max_sub_array(nums)}"

File: swift/t_code/Source/Easy/MaximumSubarray.swift
//
//  MaximumSubarray.swift
//  t_code
//
//  Created by tcui on 9/5/2018.
//  Copyright © 2018 LuckyTR. All rights reserved.
//

import Foundation

/*
 
     Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.
 
     Example:
 
     Input: [-2,1,-3,4,-1,2,1,-5,4],
     Output: 6
     Explanation: [4,-1,2,1] has the largest sum = 6.
     Follow up:
 
     If you have figured out the O(n) solution, try coding another solution using the divide and conquer approach, which is more subtle.
 
 */

class MaximumSubarray {
    
    class Solution {
        func maxSubArray(_ nums: [Int]) -> Int {
            guard !nums.isEmpty else {
                return 0
            }
            
            var sum = nums.first!
            var res = sum
            var minSum = sum
            
            for num in nums[1...] {
                sum += num
                
                let temp = max(sum - minSum, sum)
                
                if temp > res {
                    res = temp
                }
                
                if sum < minSum {
                    minSum = sum
                }
            }
            
            return res
        }
        
        func maxSubArray1(_ nums: [Int]) -> Int {
            var dp = [Int](repeating: 0, count: nums.count)
            dp[0] = nums[0]
            for i in 1..<dp.count {
                dp[i] = nums[i] + (dp[i - 1] > 0 ? dp[i - 1] : 0)
            }
            return dp.max()!
        }
    }
}
